core/trino-main/src/main/java/io/trino/execution/scheduler/TaskDescriptorStorage.java

```diff
@@ -0,0 +1,249 @@
+        updateMemoryReservation(storage.getReservedBytes());
+    }
+
+    /**
+     * Stores {@link TaskDescriptor} for a task identified by the <code>stageId</code> and <code>partitionId</code>.
+     * The <code>partitionId</code> is obtained from the {@link TaskDescriptor} by calling {@link TaskDescriptor#getPartitionId()}.
+     * If the query has been terminated the call is ignored.
+     *
+     * @throws IllegalStateException if the storage already has a task descriptor for a given task
+     */
+    public synchronized void put(StageId stageId, TaskDescriptor descriptor)
+    {
+        TaskDescriptors storage = storages.get(stageId.getQueryId());
+        if (storage == null) {
+            // query has been terminated
+            return;
+        }
+        long previousReservedBytes = storage.getReservedBytes();
+        storage.put(stageId, descriptor.getPartitionId(), descriptor);
+        long currentReservedBytes = storage.getReservedBytes();
+        long delta = currentReservedBytes - previousReservedBytes;
+        updateMemoryReservation(delta);
+    }
+
+    /**
+     * Get task descriptor
+     *
+     * @return Non empty {@link TaskDescriptor} for a task identified by the <code>stageId</code> and <code>partitionId</code>.
+     * Returns {@link Optional#empty()} if the query of a given <code>stageId</code> has been finished (e.g.: cancelled by the user or finished early).
+     * @throws java.util.NoSuchElementException if {@link TaskDescriptor} for a given task does not exist
+     */
+    public synchronized Optional<TaskDescriptor> get(StageId stageId, int partitionId)
+    {
+        TaskDescriptors storage = storages.get(stageId.getQueryId());
+        if (storage == null) {
+            // query has been terminated
+            return Optional.empty();
+        }
+        return Optional.of(storage.get(stageId, partitionId));
+    }
+
+    /**
+     * Notifies the storage that the query with a given <code>queryId</code> has been finished and the task descriptors can be safely discarded.
+     * <p>
+     * The engine may decided to destroy the storage while the scheduling is still in process (for example if query was cancelled). Under such
+     * circumstances the implementation will ignore future calls to {@link #put(StageId, TaskDescriptor)} and return
+     * {@link Optional#empty()} from {@link #get(StageId, int)}. The scheduler is expected to handle this condition appropriately.
+     */
+    public synchronized void destroy(QueryId queryId)
+    {
+        TaskDescriptors storage = storages.remove(queryId);
+        if (storage != null) {
+            updateMemoryReservation(-storage.getReservedBytes());
+        }
+    }
+
+    private synchronized void updateMemoryReservation(long delta)
+    {
+        reservedBytes += delta;
+        while (reservedBytes > maxTaskDescriptorStorageMemoryInBytes) {
+            // drop a query that uses the most storage
```

Comment:
I am not sure killing the biggest query is best approach. The queries which use the most memory are the biggest, and probably already executing for a long time. Killing such query is costly. I would rather kill the queries which made least progress already (newest ones).

Reply:
> I am not sure killing the biggest query is best approach. The queries which use the most memory are the biggest, and probably already executing for a long time. Killing such query is costly. I would rather kill the queries which made least progress already (newest ones).

I was also thinking about this. Contrary killing any other query than the one that has the most meta information buffered might be confusing. It is easier to communicate to a user that their query got killed because it scans very large tables or tables with a lot of small files (that results in high memory utilization in split enumeration) vs communicating that their queries are killed because there's other query running in a cluster that scans large tables.

Regardless, it still just a stop gap approach. If we see that the queries are killed because of this limitation and it is problematic we will invest into spilling to prevent failures of this kind altogether. 

> Also I am not a big fan of entangling storage and killing logic. But I must agree it makes interfaces simpler.

Killing is just a temporary solution. The end goal is to have spilling capabilities implemented in the storage. Thus I'm not sure if it is worth overthinking how the queries are killed, as eventually we don't want the queries to be killed at all.